rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,697 @@
+```
+
+And you get back a fully fledged file system fetched from the repo. As the SDK is focused on running on context of browser.
+
+Above function call can do a fetch of just enough resources to do further actions.
+
+`git clone --depth 1 --sparse --no-checkout --filter=blob:none https://github.com/inlang/inlang`
+
+After this an app would most likely want to fetch a file to edit. It can do:
+
+```js
+// checkout just this file and put it in the fs (using sparse-checkout)
+// it returns a new inlang, so 'fs' below works and is filled with correct file system
+// again this can be a signal instead or hook, not just a variable
+inlang = await inlang.checkoutFile("inlang.config.js")
+
+// `.fs` is one way you can get access to the file system
+// readFile returns a string representation of the file so you can modify it
+// need to think about how fs. would work
+// maybe there is way to get access to fs in a better way
+let inlangConfig = await inlang.fs.readFile("inlang.config.js")
+// it would be great if `inlang` type definition would update on 'checkout', 'clone'
+// so typescript can complain if 'inlang.config.js' file is not there
+
+// then you can make edits to inlangConfig string that
+// it could be the case that inlangConfig not a normal variable
+// but a signal returned (to work with solid) or hook (for react)
+
+// where inlangConfig is the modified version
+// aware there are most likely issues with just editing file like this
+// i'm sure there is a better way more streamlined way to make an edit to something and
+// commit it to fs after
+// this is just one example of how it can happen
+inlangGit.overwriteFile("inlang.config.js", inlangConfig)
+
+// Then git add the file
+inlangGit.add("inlang.config.js")
+
+// Commit
+inlangGit.commit("inlang.config.js: update readResources")
+
+// And git push
+const conflicts = await inlangGit.push("inlang.config.js")
+
+// Can also get conflicts on pull
+const conflicts = await inlangGit.pull()
+
+// you try fix conflicts and
+inlangGit.push() // until it succeeds
+```
+
+Sparse-checkout under the hood does this (for reference):
+
+```
+git config core.sparseCheckout true
+echo "inlang.config.js" >> .git/info/sparse-checkout
+git checkout
+```
+
+The API for above is quite readable and nice. And powerful too in some ways. As you don't have to think about how fetch happens. It's done for you automatically.
+
+```js
+const inlang = await gitSDK.clone("https://github.com/inlang/inlang.git", { depth: 1 })
+```
+
+You can also as second arg to `clone` send some options to already fetch up to certain depth of commits or any other option that would make sense to run for Git to be run in the browser.
+
+In many ways this opens up the API surface too and lets us explore what to expose to users and what not. A lot of complexity can be hidden away. Namely all the memfs api learning you will need to do, together with more lines of code and potential bugs.
```

Comment:
I don't see a learning requirement with filesystems.

Yes, we use `memfs` _in the browser_ right now. But there is no learning. Just call `readFile()` as one would do in a local environment like Node.js.  

Related to a remark earlier, abstracting the filesystem away makes the git-sdk less versatile for no immediate upside other than saving one from passing the filesystem as arguments to git commands like `clone(fs)`.